### DISCUSSION TITLE
Script, catch switch state and show on webui

### DISCUSSION THREAD

**Discussion by rice103** [arendst/Tasmota#10359] [Q&A] [answered]
It's a simple request but i cannot catch the switch value in no way.
Switch 3 it's for me state of a PIR and first problem being make no action for that switch. So i connecting to empty pin by power3.
In this way i can raise the event:
```
>E
sw=sw[3]
if upd[sw]>0
then 
=>print switch updated %sw%
=>Publish %topic%/stat/PIR %sw%
endif
```
On console i can see the print and the publish
but **sw** it's always 0.00.......
Do you know why?

**Comment by Frogmore42** [comment]
Really not enough information to help you.  Look at the documentation on switches or you could use rules, I suppose. https://tasmota.github.io/docs/Rules/

Most likely you should use PIR as a button, but it depends on the PIR.  Some give a short pulse, use button for that and set the topic.  This is the old school way of doing it.

If the PIR output stays active for a long time, you should use Switch.  But, it sounds like your PIR is not staying active for long when it senses motion.

**Reply by barbudor** [reply]
@Frogmore42 PIR must be configured as switches. The standard behavior of button is not good for PIR.

**Comment by gemu2015** [comment]
if you look at the documentation sw[x] index starts from 0 not from 1.
this was initially a mistake since other indices go from 1 to N but since a couple of users already used index 0 i did not change this.

Events (section >E) are executed when Tasmota generates a JSON EVENT/ RESULT. (can be seen in console)

since it is not really relevant in your application you can also watch for a change in section >F (every 100ms)

**Reply by rice103** [reply]
I tried with sw[2] but sw[3] and it's the same...
Now i'm tring with F section

**Comment by rice103** [comment]
Look this:
https://tasmota.github.io/docs/PIR-Motion-Sensors/
There it's described how configure tasmota for send a custom mqtt message on change of switch state but with rules but i can't do that with script, that's my problem

**Comment by gemu2015** [comment]
ok, this used to work 1 year ago. there have been many changes to tasmota switches since that time.
sw[x] reports switch last state which definitely should change.

did you try with button and    (PUSH_IGNORE) ?

code:
     case PUSH_IGNORE:
          Switch.last_state[i] = button;                        // Update switch state before publishing
          MqttPublishSensor();

i myself do nearly all that stuff inside scripts itself also PIR.
e.g.
GPIO12 in Tasmota to "none"

\>D
pir=0

\>B
; set pin 12 to Input with Pullup
spinm(12 P)

\>F
pir=pin(12)
if chg[pir]>0
then
...

**Comment by gemu2015** [comment]
i will evaluate this sw[x] problem and update the source soon

**Comment by gemu2015** [comment]
ok works like expected:

configure switch1 in Tasmota on gpio x
decouple from relay setoption114

\>E
print event %sw[0]%

**Reply by rice103** [reply]
if i decouple with setoption114 event will not raised more

**Comment by barbudor** [comment]
@rice103 as stated on Discord, apart from the index issue, you should use 'chg' and not 'upd'
See the 1st example in the cookbook

**Reply by rice103** [reply]
but if i use chg i don't detect the change while if i use upd yes

**Reply by barbudor** [reply]
@rice103 You have to decide if you want this conversation to be on Discord or Github
Not going to dupplicate my answers

**Comment by gemu2015** [comment]
ha, discovered the bug!
you may not call a variable sw !!
i called it pir and it worked.
when i used sw it did not work.

must be a name clash with sw[x]

**Reply by barbudor** [reply]
Hum, this is from the Tasmota doc itself:
```
; check if switch changed state
sw=sw[1]
if chg[sw]>0
then =>power1 %sw%
endif
```
Should it be updated ?

**Reply by gemu2015** [reply]
yes it worked at that time, but not now.

**Reply by gemu2015** [reply]
i will investigate where the bug is, and if easy to remove will do, else i update the docs

**Reply by gemu2015** [reply]
bug is fixed, will do a pr soon

**Reply by gemu2015** [reply]
merged

**Comment by rice103** [comment]
Ok thanks i'll try.
Now testing i found another solution reading directly pin:

```
>F
if pin[16]==0
then sw=0
endif
if pin[16]==1
then sw=1
endif

if sw!=oldSw
then 
=>printswitch changed %sw%
=>Publish %topic%/stat/PIR %sw%
endif
oldSw=sw
```

**Reply by rice103** [reply]
2 comment:
1) i cannot use chg[] because don't work
2) i cannot assign directly pin[16] to sw variable but i must write two if

**Reply by rice103** [reply]
another comment:
3) i'm using pin[16] 
pin[1..16] not pin[0..15]

**Comment by gemu2015** [comment]
please read my comment above!

do NOT use a variable named sw !

it all works if you use e.g. pir!

**Reply by rice103** [reply]
Ok i'm tring but it's only for make point of situation

**Reply by rice103** [reply]
Result:
11:47:14.378 MQT: home.closet.shutter/cmnd/POWER3 = ON
11:47:14.429 pin[16] 1.00
11:47:14.432 sw[2] 0.00???
11:47:14.434 sw[3] 0.00???
11:47:14.443 MQT: home.closet.shutter/stat/RESULT = {"POWER3":"ON"}
11:47:14.447 MQT: home.closet.shutter/stat/POWER3 = ON

script:
```
>E
print pin[16] %pin[16]%
print sw[2] %sw[2]%
print sw[3] %sw[3]%
```

**Reply by rice103** [reply]
Don't work

**Reply by rice103** [reply]
YES WORK!

**Reply by rice103** [reply]
```
11:54:21.747 pin[16] 0.00
11:54:21.750 sw[2] 0.00
11:54:21.753 sw[3] 1.00
```

**Comment by gemu2015** [comment]
it looks like you have still sw=0 in >D ?

**Reply by rice103** [reply]
sorry you're right!!!!
but i must use sw[3] not sw[2]

**Comment by rice103** [comment]
Remains only one problem:
if I use SetOption114 1 the events are not raised
but if I don't use it and don't assign power3 I get a "command error" back on console
Is it possible to avoid assigning power3 but keep raising events?

**Comment by gemu2015** [comment, marked as answer]
it works without problems in my setup

\>D
pir=0
\>B
->setoption114 1
->switchmode 1

\>E
pir=sw[0]
if chg[pir]>0
then
print event %pir%
endif

\>S
print %sw[0]%

**Reply by rice103** [reply]
how is module configuration?
